Tasks/SonarQubePostTest/PRCA/Orchestrator.ps1

```diff
@@ -1,18 +1,113 @@
-
 . $PSScriptRoot/ReportProcessor.ps1
+. $PSScriptRoot/IssuesProcessor.ps1
 
 function HandleCodeAnalysisReporting
 {	
     Write-Verbose "HandleCodeAnalysisReporting..."
     if (IsPrBuild)
     {   
-        Write-Host "Fetching code analysis issues and posting them to the PR..." 
+        Write-Host "Fetching code analysis issues and posting them to the PR..."        
         $newIssues = FetchAnnotatedNewIssues                
         
-        if ($newIssues.Length -eq 0)
-        {            
+        if ( ($newIssues -eq $null) -or ($newIssues.Length -eq 0))
+        {
             Write-Host "The SonarQube analysis did not find any new issues."
-            return            
+            return
         }
+
+        LoadClientOMAssemblies 
+        PostIssuesToCodeReview $newIssues   
     }	
 }
+
+#region Private 
+
+function PostIssuesToCodeReview
+{       
+    param ([string]$newIssues)     
+    
+    $vssConnection = [Microsoft.TeamFoundation.DistributedTask.Task.Internal.Core.TaskContextHelper]::GetVssConnection($distributedTaskContext); 
+    Assert ( $vssConnection -ne $null ) "Internal error: could not retrieve the VssConnection object"
+        
+    $gitClient = $vssConnection.GetClient("Microsoft.TeamFoundation.SourceControl.WebApi.GitHttpClient")            
+    $discussionClient = $vssConnection.GetClient("Microsoft.VisualStudio.Services.CodeReview.Discussion.WebApi.DiscussionHttpClient")    
+    $codeReviewClient = $vssConnection.GetClient("Microsoft.VisualStudio.Services.CodeReview.WebApi.CodeReviewHttpClient") # ???
+        
+    Assert ( $gitClient -ne $null ) "Internal error: could not retrieve the GitHttpClient object"
+    Assert ( $discussionClient -ne $null ) "Internal error: could not retrieve the DiscussionHttpClient object"
+    Assert ( $codeReviewClient -ne $null ) "Internal error: could not retrieve the CodeReviewHttpClient object"
+             
+    $sourceBranch =  GetTaskContextVariable "Build.SourceBranch"
+    $pullRequestId = RetrievePullRequestId $sourceBranch
+    Write-Verbose  "Pull Request Id $pullRequestId"
+    
+    $repositoryId = GetTaskContextVariable "build.repository.id"    
+    $teamProject = GetTaskContextVariable "system.teamProject"
+
+    Assert (![String]::IsNullOrWhiteSpace($repositoryId)) "Internal error: could not determine the build.repository.id"
+    Assert (![String]::IsNullOrWhiteSpace($teamProject)) "Internal error: could not determine the system.teamProject"   
+
+    $pullRequest = $gitClient.GetPullRequestAsync($project, $repositoryId, $pullRequestId).Result;
+    $artifactUri = GetArtifactUri $pullRequest.CodeReviewId $pullRequest.Repository.ProjectReference.Id $pullRequestId
+           
+    ProcessIssues $newIssues $codeReviewClient $gitClient $discussionClient $pullRequest $teamProject $repositoryId $sourceBranch $artifactUri        
```

Comment:
Done!

From: Eyal Lantzman [mailto:notifications@github.com]
Sent: Tuesday, March 1, 2016 3:27 PM
To: Microsoft/vso-agent-tasks <vso-agent-tasks@noreply.github.com>
Cc: Bogdan Gavril <bgavril@microsoft.com>
Subject: Re: [vso-agent-tasks] Have the orchestrator new up all the actors needed to publish issues (#1319)


In Tasks/SonarQubePostTest/PRCA/Orchestrator.ps1<https://github.com/Microsoft/vso-agent-tasks/pull/1319#discussion_r54582348>:

> +    Assert ( $codeReviewClient -ne $null ) "Internal error: could not retrieve the CodeReviewHttpClient object"

> +

> +    $sourceBranch =  GetTaskContextVariable "Build.SourceBranch"

> +    $pullRequestId = RetrievePullRequestId $sourceBranch

> +    Write-Verbose  "Pull Request Id $pullRequestId"

> +

> +    $repositoryId = GetTaskContextVariable "build.repository.id"

> +    $teamProject = GetTaskContextVariable "system.teamProject"

> +

> +    Assert (![String]::IsNullOrWhiteSpace($repositoryId)) "Internal error: could not determine the build.repository.id"

> +    Assert (![String]::IsNullOrWhiteSpace($teamProject)) "Internal error: could not determine the system.teamProject"

> +

> +    $pullRequest = $gitClient.GetPullRequestAsync($project, $repositoryId, $pullRequestId).Result;

> +    $artifactUri = GetArtifactUri $pullRequest.CodeReviewId $pullRequest.Repository.ProjectReference.Id $pullRequestId

> +

> +    ProcessIssues $newIssues $codeReviewClient $gitClient $discussionClient $pullRequest $teamProject $repositoryId $sourceBranch $artifactUri

Scratch that - probably better as private fields set during initialization (which you'll need to extract out of this method)

—
Reply to this email directly or view it on GitHub<https://github.com/Microsoft/vso-agent-tasks/pull/1319/files#r54582348>.